### PR TITLE
Fix interaction with the castle bridge for flying units

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
In the original game, flying unit could not land on the gates cell, even if gates are already open or destroyed:

OG

At the moment, in fheroes2 flying units don't interact with gates and bridge in any way, what leads to situations like this:

Phoenixes fail

In this PR I propose some changes that will cause flying units to interact with the bridge like everyone else - the bridge should be lowered first to allow landing on the gates cell:

Phoenixes OK

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Sorry for multiple force pushing, it seems that "scan-build analyzer" workflow constantly fails on apt-get update step.

**Comment by ihhub** [comment]
> Sorry for multiple force pushing, it seems that "scan-build analyzer" workflow constantly fails on apt-get update step.

Hi @oleg-derevenetz , it's totally fine to have any reasonable number of pushes or commits. We do not restrict it.

File: src/fheroes2/battle/battle_board.h [hunk 1]
```diff
@@ -73,8 +73,8 @@ namespace Battle
         static bool isNearIndexes( s32, s32 );
         static bool isValidIndex( s32 );
         static bool isCastleIndex( s32 );
-        static bool isMoatIndex( s32, int );
-        static bool isBridgeIndex( s32, int );
+        static bool isMoatIndex( s32, const Unit & );
+        static bool isBridgeIndex( s32, const Unit & );
```
Comment (ihhub) [inline src/fheroes2/battle/battle_board.h:77]: Could you please add input argument names? The old code is filled with this weird coding style. We shouldn't do this as such approach makes the code less readable.

File: src/fheroes2/battle/battle_bridge.cpp [hunk 2]
```diff
@@ -68,41 +68,45 @@ bool Battle::Bridge::isBridgeOccupied( void ) const
...
-        if ( ( prevPos == BELOW_BRIDGE_CELL || prevPos == ABOVE_BRIDGE_CELL ) && b.GetColor() == Arena::GetCastle()->GetColor() )
+    if ( b.isFlying() ) {
+        if ( dstPos == GATES_CELL ) {
             return true;
+        }
     }
-    else if ( dstPos == MOAT_CELL ) {
-        if ( prevPos != GATES_CELL && b.GetColor() == Arena::GetCastle()->GetColor() )
+    else {
+        const s32 prevPos = b.GetHeadIndex();
```
Comment (ihhub) [inline src/fheroes2/battle/battle_bridge.cpp:81]: Could you please replace `s32` by **int32_t**? We would like to remove this `using` cases from the code (slowly).

File: src/fheroes2/battle/battle_bridge.cpp [hunk 3]
```diff
@@ -68,41 +68,45 @@ bool Battle::Bridge::isBridgeOccupied( void ) const
...
         return false;
 
-    const s32 prevPos = b.GetHeadIndex();
-
-    if ( dstPos == GATES_CELL ) {
-        if ( prevPos == CELL_AFTER_GATES )
-            return true;
-        if ( ( prevPos == BELOW_BRIDGE_CELL || prevPos == ABOVE_BRIDGE_CELL ) && b.GetColor() == Arena::GetCastle()->GetColor() )
+    if ( b.isFlying() ) {
+        if ( dstPos == GATES_CELL ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_bridge.cpp:76]: We could do like this: `return dstPos == GATES_CELL;`. So it's just one-two CPU ticks faster but... :)

File: src/fheroes2/battle/battle_board.cpp [hunk 4]
```diff
@@ -184,9 +184,14 @@ void Battle::Board::SetScanPassability( const Unit & unit )
     at( unit.GetHeadIndex() ).SetDirection( CENTER );
 
     if ( unit.isFlying() ) {
-        for ( iterator it = begin(); it != end(); ++it )
-            if ( ( *it ).isPassable3( unit, false ) )
+        const Bridge * bridge = Arena::GetBridge();
+        const bool isPassableBridge = bridge == nullptr || bridge->isPassable( unit );
+
+        for ( iterator it = begin(); it != end(); ++it ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_board.cpp:190]: Please use here normal `size_t` based loop. The compiler will optimize this and `it - begin()` will be just `size_t` value. It also would be a bit more readable. However, I'm leaving it up to you to decide.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @oleg-derevenetz , I left just few comments here. Could you please take a look? Also I requested @Branikolog to test this version.

File: src/fheroes2/battle/battle_board.h (hunk 1 above)
Comment (oleg-derevenetz) [reply src/fheroes2/battle/battle_board.h:77]: Of course. Just did it "in the style of the surrounding code" :) I'll fix this.

**Comment by Branikolog** [comment]
I made a few tests... Everything seems to work fine.

**Comment by ihhub** [comment]
@oleg-derevenetz , huge thanks for fixing this issue!

**Pull request merged**